Content.Server/_DV/Autoclave/AutoclaveSystem.cs

```diff
@@ -0,0 +1,105 @@
+using Content.Server.Power.EntitySystems;
+using Content.Shared._DV.Autoclave;
+using Content.Shared._DV.Surgery;
+using Content.Shared.Power;
+using Content.Shared.Storage.Components;
+using Content.Shared.Storage.EntitySystems;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Timing;
+using System.Linq;
+
+namespace Content.Server._DV.Autoclave;
+
+public sealed class AutoclaveSystem : EntitySystem
+{
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly PowerReceiverSystem _power = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+    [Dependency] private readonly SharedEntityStorageSystem _entityStorage = default!;
+    [Dependency] private readonly SharedSurgeryCleanSystem _surgeryClean = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<AutoclaveComponent, PowerChangedEvent>(OnPowerChanged);
+        SubscribeLocalEvent<AutoclaveComponent, StorageAfterOpenEvent>(OnOpened);
+        SubscribeLocalEvent<AutoclaveComponent, StorageAfterCloseEvent>(OnClosed);
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<AutoclaveComponent, SurgeryCleansDirtComponent>();
+
+        while (query.MoveNext(out var uid, out var comp, out var cleansDirt))
+        {
+            if (_gameTiming.CurTime < comp.NextUpdate)
+                continue;
+
+            comp.NextUpdate += comp.UpdateInterval;
+
+            var isPowered = _power.IsPowered(uid);
+            var isClosed = !_entityStorage.IsOpen(uid);
+
+            if (!(isPowered && isClosed))
+                continue;
+
+            SharedEntityStorageComponent? storageComponent = null;
+            if (!_entityStorage.ResolveStorage(uid, ref storageComponent))
+                continue;
+
+            foreach (var containedEntity in storageComponent.Contents.ContainedEntities)
+            {
+                _surgeryClean.DoClean(new(uid, cleansDirt), containedEntity);
+            }
+
+            var hasDirtyContents = storageComponent.Contents.ContainedEntities.Any(contained => _surgeryClean.RequiresCleaning(contained));
+            UpdateVisuals(uid, true, true, hasDirtyContents);
+        }
+    }
+
+    private void UpdateVisuals(EntityUid ent, bool isPowered, bool isClosed, bool hasDirtyContents)
+    {
+        var (greenLight, redLight) = (isPowered, isClosed, hasDirtyContents) switch
+        {
+            (false, _, _) => (false, false),
+            (true, false, _) => (false, true),
+            (true, true, true) => (true, false),
+            (true, true, false) => (false, true),
+        };
+        _appearance.SetData(ent, AutoclaveVisuals.IsProcessing, greenLight);
+        _appearance.SetData(ent, AutoclaveVisuals.IsIdle, redLight);
+    }
```

Comment:
hasDirtyContents is always calculated for this call so just move the 2 occurences to this method to make it better